### PR TITLE
Added the random string reference to the letter

### PR DESCRIPTION
**Pull request by minglis**
- uses the reference field on the notifications table to store a 16char random string used to cross reference DVLA letters back to the notification
- used as letter barcode does not have space for a UUID notification id

- [x] Depends on [https://github.com/alphagov/notifications-utils/pull/149](https://github.com/alphagov/notifications-utils/pull/149)

Renamed the numeric_id to notification_reference in utils and changed validation rules to match this

Note also the persist_notification method set "reference" to be "client_reference" which is confusing and they are different things, so fixed this too.

### CODE REVIEW

**Comment by minglis** [comment]
the existing test_create_dvla_file_contents test in test_tasks is a little confusing, it mocks the template from utils and then checks the args, passed to it. I change the test a bit to check for both references (one from each created notification) but it threw me a bit so whoever looks at this could you look at it too and see what the thoughts are. ta.

File: app/v2/notifications/post_notifications.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def post_notification(notification_type):
                                         notification_type=notification_type,
                                         api_key_id=api_user.id,
                                         key_type=api_user.key_type,
-                                        reference=form.get('reference', None),
+                                        client_reference=form.get('reference', None),
```
Comment (leohemsted) [inline app/v2/notifications/post_notifications.py]: if someone posted an email with a reference, and then we send it to SES, did we overwrite the user's reference with SES's one?

Comment (leohemsted) [inline app/v2/notifications/post_notifications.py]: do we want/need/have the ability to migrate old data to client_reference?

Comment (minglis) [inline app/v2/notifications/post_notifications.py]: Actually I don't think so. The persist only ever set the client-reference. Even if that was from the "reference" post body field.

So the data would not be overwritten.

Comment (minglis) [inline app/v2/notifications/post_notifications.py]: my change adds in the reference method param, and puts the two things back into the correct params.

Comment (leohemsted) [inline app/v2/notifications/post_notifications.py]: ah right, just params, not actual db values. gotcha. :ok_man:

**Pull request merged**